profileがセットされたら、そのprofile使って再advertiseする他

tsuu32 opened the pull request:
### やったこと
- （リスナー限定機能）DJを見失ったらDJのアイコンと名前が薄くなる
  - そのためにconnectedDJを(peerID: MCPeerID, state: MCSessionState)?型にする
  - connectedDJ.stateでDJと接続してるかを保持する
  - isDJがfalseのときは絶対connectedDJはnilではないという信念のもとのconnectedDJ!
- profileがセットされたら、そのprofileを使って再advertiseする
  - Twitterのログイン情報がUserDefaultsに保存されていてTwitterのprofileが有効なら、自動的にそれを使ってadvertiseするようになった
- 誰かと接続したときだけでなく、誰かと切断したときもSessionタブの中身を更新

### バグっぽいところ
DJを見失ってDJのアイコンと名前が薄くなるとき、DJの名前がデバイス名になってしまうことがある
`updateMembers`の`DispatchQueue`周りがアレそう （関連 #91）

### やるかどうか迷ってるところ
DefaultsControllerのprofileをPeerProfile型にして、初期値にPeerProfile(name: UIDevice.current.name, imageUrl: nil)を与えるようにするやつ

### やってほしいこと
動作確認。（リスナからの曲のリクエストとかがちゃんと動いてるか確認してほしい）

tsuu32 committed:
connectedDJを(peerID: MCPeerID, state: MCSessionState)?型に変更

リスナがDJと切断した（見失った）ときstateに.notConnectedが入る
逆にDJと接続中はstateに.connectedが入る

isDJがfalseの中ならconnectedDJは絶対にnilでないはずなのでforce unwrapする（最悪コード）

tsuu32 committed:
リスナがDJを見失ったとき、DJのアイコンと名前のalphaを小さくする

tsuu32 committed:
設定でprofileが変更されたらそのprofileを使ってadvertiseを再開する

技術的な制限:
DJはSettingのTwitterのSwitchを押すたびにadvertiseし直している
Join as a listener SceneでDJの再advertiseを検知してtable viewを更新させたい
しかし、MCNearbyServiceBrowserDelegateのメソッドが呼ばれないため、新しいadvertiseに乗っている情報を得ることができない
よってDJの設定変更のタイミングでJoin as a listener Sceneのtable viewをリアルタイムに更新することはできない

tsuu32 committed:
WelcomeViewController.swiftですでに不要なisViewAppearedAtLeastOnce変数を削除

tsuu32 committed:
リスナーがDJを見失ったときだけでなく、どのピアも誰かと切断したらupdateMembersする

あと、よく考えたら不要な通知を削除

amylaseF85 commented:
グレーアウトしてるときにConnectingだとおかしいので、修正願いたい
![IMG_B5721246278F-1](https://user-images.githubusercontent.com/37996341/71388809-3870d280-263d-11ea-96ea-6614ade2bb71.jpeg)

MemberSceneの「Connecting」ってlabelをMemberViewController.swiftで呼び出して

amylaseF85 commented:
子機からリクエスト飛ばない...
OSのせいかもしれない（親：13.1.3、子：13.3）

親にアップデートかけて再試行する

tsuu32 commented:
@amylaseF85
子機からリクエスト飛ばせないのは、俺の書いたコードがバグってました

amylaseF85 commented:
やっと言ってる意味わかったわ

amylaseF85 commented:
==にしたらちゃんと子機からリクエストが届きました
動作に関しては問題なさそう

amylaseF85 commented on DJYusaku/WelcomeViewController.swift:
```diff
@@ -13,8 +13,6 @@ class WelcomeViewController: UIViewController {
     
     @IBOutlet weak var doneButtonItem: UIBarButtonItem!
     
-    static private var isViewAppearedAtLeastOnce: Bool = false;
-    
```
これなんで消えてるんだっけ
なんか言ってた気がするけどいまいち思い出せない

yaplus replied:
#75 でWelcomeViewのDone表示・非表示の切り替え実装を変えたからだと思う

amylaseF85 replied:
なるほどね、プルリク本体になかったので

yaplus commented:
Connecting表記をMissingとかにしてほしい感あるわね

amylaseF85 commented:
> Connecting表記をMissingとかにしてほしい感あるわね

むしろきっぱりDisconnect(ed)でもいいと思ってる

yaplus commented:
Disconnectedはユーザーが能動的にDisconnectした感がある  
アプリの側が見失った的なニュアンスでMissingにしましょう

yaplus commented:
というかMemberView全体的に殺風景でダサいから後で作り直す

tsuu32 committed:
最近の変更でリスナがDJにリクエストを送らなくなっていたのを修正

間違えてた

tsuu32 committed:
リスナがDJを見失ったときMissingにする

tsuu32 committed:
最近の変更で子機がadvertiseしてしまっていたのを修正

tsuu32 commented:
ちょっと色々バグだらけなのでリファクタリングします
具体的には、

DefaultsControllerのprofileをPeerProfile型にして、初期値にPeerProfile(name: UIDevice.current.name, imageUrl: nil)を与えるようにするやつ

ConnectionControllerのisDJを消して、isDJメソッドにする（内容はreturn connectedDJ ==nil）

yaplus commented:
@tsuu32 

PeerProfileの件了解です。

isDJについてはコンピューテッドプロパティ（関数みたいな変数）使いましょう。

```swift
    private(set) var isDJ: Bool {
        get {
            return self.connectedDJ ==nil
        }
    }
```

もしかしたら `private(set) ` いらんかもしれん

yaplus commented:
というかDefaultsControllerが持ってる変数も参照の都度UserDefaultsから引っ張ってくるコンピューテッドプロパティにしてもいいのかもしれん  
今ちゃんと動いてるから下手に直さなくていいと思うけど…

tsuu32 committed:
リスナーが一度アプリを離れて10秒ほど経ってアプリに戻ったとき、再接続できていなかった不具合を修正

原因:
リスナーがアプリを離れて10秒ほど経つと接続が切断する。（バックグラウンドのためリスナーで.notConnectedなsession(_ session: MCSession, peer peerID: MCPeerID, didChange state: MCSessionState)は呼ばれていない）
アプリ復帰直後にConnectionControllerのhandleViewWillEnterForegroundが呼ばれて親機に接続しにいく。
しかし、その後に.notConnectedなsession(_ session: MCSession, peer peerID: MCPeerID, didChange state: MCSessionState)が呼ばれて、
self.connectedDJ!.state = .notConnectedとなってしまっていた。

解決方法:
リスナーはアプリを離れるときsession.disconnect()する（ただしconnectedDJにnilは入れない）

tsuu32 commented:
このプルリクにただでさえいろいろ詰め込まれてるのにさらに詰め込むのはアレかと思ったので「DefaultsControllerのprofile変えるのとConnectionControllerのisDJ変えるの」は別のブランチでやりたい
とりあえず現在のでレビューしてほしいです

tsuu32 committed:
Join as a listener SceneのDj Nameラベルの縦幅を広くする

jとかgとかyみたいなベースラインより下にはみ出る文字で表示が切れてる問題を修正

yaplus reviewed (approved):
動作はだいたい大丈夫そう  
たまに端末名になるやつはなんとかしたいね  
（PeerProfile?をPeerProfileにしたら治ったりする？）

amylaseF85 commented:
まーじします

amylaseF85 commented:
おしました
